### PR TITLE
Implement Product Page V2 flag behavior

### LINKED ISSUE #24001: [Add support for experimental features] Implement Product Page V2 flag behavior
This issue is an item in the [Add support for experimental features](https://github.com/PrestaShop/PrestaShop/issues/23235) project

You can read about the project [here](https://github.com/PrestaShop/PrestaShop/issues/23235).

Following the introduction of the [Product Page V2 flag](https://github.com/PrestaShop/PrestaShop/issues/23999) BO users will be able to enable/disable the flag from the Back Office.

**We need to make sure that, when the flag is disabled, the Product Page BO page V2 is not available.**

**We need to make sure that, when the flag is enabled, the Product Page BO page V2 is available from multiple CTA.**

### Cannot access v2 Page when flag is disabled

Scenario: Cannot see 'New product on experimental page' button in Product Catalog
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is disabled
THEN on the BO listing page 'Product Catalog' I cannot see the 'New product on experimental page' button

Scenario: Cannot see 'Edit on experimental page' button in 'Edit product' v1 page
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is disabled
WHEN I open a 'Edit Product' BO page
THEN I cannot see the 'Edit on experimental page' button

Scenario: Redirect user from v2 page to v1 page
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is disabled
WHEN I attempt to visit the BO page v2 using the URL `/admin-dev/index.php/sell/catalog/products-v2/{PRODUCT-ID}/edit?_token={TOKEN}`
THEN I am redirected to BO page v1
AND a flash message tells me why I was redirected

**Screenshot**
Capture d’écran 2021-04-13 à 23 01 31

### Can access v2 Page when flag is enabled

Scenario: Can see 'New product on experimental page'  button in Product Catalog
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is enabled
THEN on the BO listing page 'Product Catalog' I can see the 'New product on experimental page'  button

**Screenshot**
Capture d’écran 2021-04-14 à 16 15 59

Scenario: Can see 'Edit on experimental page' button in 'Edit a product' v1 page
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is enabled
WHEN I open a 'Edit product' BO page
THEN I can see the 'Edit on experimental page' button

Scenario: Can see 'Edit on experimental page' button in Product Catalog
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is enabled
WHEN I open the Product Catalog BO page and I click on the `dots menu` on the right of a product
THEN I can see a link to Page v2 with name 'Edit on experimental page'

**Screenshot**
Capture d’écran 2021-04-13 à 22 28 54

Scenario: Can see 'Back to standard page' button in Product Catalog
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is enabled
WHEN I open a Product Page V2 page
THEN I can see a link to Page v1 with name 'Back to standard page'

### PR DESCRIPTION
**Pull request by matks**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/24001
| How to test?      | See below
| Possible impacts? | Product page buttons only

### Description

In this PR I introduce and rename multiple buttons to navigate between Product current pages (= V1) and experimental Product Page (= V2). I added a redirection from V2 page to V1 page when user attempts to reach V2 page although the flag is disabled.

### How to test

1. Install PS using this branch
2. If the Feature Flag for Product Page V2 is not loaded*, load it in database using SQL script
```
INSERT INTO `ps_feature_flag` (`name`, `state`, `label_wording`, `label_domain`, `description_wording`, `description_domain`)
VALUES
   ('product_page_v2', 0, 'Experimental product page', 'Admin.Advparameters.Feature', 'This page is a work in progress. It includes new combination management features and other features under development (virtual products, packs, etc.)', 'Admin.Advparameters.Help');
   ```
3. Visit BO page Configure > Adv Parameters > Experimental Features to enable or disable the flag.
4. Follow the stories written in https://github.com/PrestaShop/PrestaShop/issues/24001

*It should be loaded because https://github.com/PrestaShop/PrestaShop/pull/24037 was merged

#### Edit button in listing

Capture d’écran 2021-04-13 à 22 28 54

<hr>

#### Add new button in listing

Capture d’écran 2021-04-14 à 16 15 59

<hr>

#### Edit button using experimental page in Page V1

Capture d’écran 2021-04-14 à 16 22 58

<hr>

#### Back to V1 page in experimental page

Capture d’écran 2021-04-14 à 16 23 10

<hr>

#### Redirection message

Capture d’écran 2021-04-13 à 23 01 31

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24038)

### CODE REVIEW

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]:
```suggestion
        if (!$this->isProductPageV2Enabled()) {
```

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]:
```suggestion
        if (!$this->isProductPageV2Enabled()) {
```

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]: Move to use section

File: src/PrestaShopBundle/Resources/views/Admin/Product/ProductPage/Blocks/footer.html.twig
Comment (Progi1984) [inline src/PrestaShopBundle/Resources/views/Admin/Product/ProductPage/Blocks/footer.html.twig]:
```suggestion
      {% if isProductPageV2Enabled == true %}
```

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]:
```suggestion
                [
                    sprintf('<a href="%s">', $this->get('router')->generate('admin_feature_flags_index')), 
                    '</a>'
                ]
```

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]:
```suggestion
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (Progi1984) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]: Ping @matks and it's ok ;)

Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]: Too fast for me 😛  look at https://github.com/PrestaShop/PrestaShop/pull/24038/commits/392e403b7904bd000c125010415b18f7b9642b6f

**Review by Progi1984 — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by Julievrz — DISMISSED** [review]
LGTM, thank you! @matks

File: src/PrestaShopBundle/Controller/Admin/ProductController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/ProductController.php]: I'm not sure this const should be in this controller, nor in the new one Because they become co-dependent just to get a feature flag name You could simply use a hardcoded value since it's the ID/slug/name a dynamic data in the database (a bit like we do with Configuration keys)

If you really want to use a const as a reference (which I agree is cleaner) it should be in an independent object that is related to FeatureFlag, not the controller This will avoid co-dependency and besides the `product v2` feature is not only limited to one controller (it's composed of several ones, plus CQRS command, plus the list, ...)

So I'd suggest creating either a generic `FeatureFlagSettings` class to hold such const, either a generic one which would include future other flags along time or a specific one `ProductFeatureFlagSettings` which is dedicated to product

**Review by jolelievre — COMMENTED** [review]
Everything is good to me, except the const that shouldn't be placed here but in a separate class

File: src/PrestaShopBundle/Controller/Admin/ProductController.php
Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/ProductController.php]: Fixed https://github.com/PrestaShop/PrestaShop/pull/24038/commits/feb0b3807ead2c1f0757aadae6a4c3ca5fc4988f

**Review by jolelievre — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hello @matks !

Here is the recap of the test of the PR. All these behaviors were tested in Desktop and Mobile display, and in standard and Multistore configuration 

Scenario: Cannot see 'New product on experimental page' button in Product Catalog
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is disabled
THEN on the BO listing page 'Product Catalog' I cannot see the 'New product on experimental page' button
=> :heavy_check_mark: 

Scenario: Cannot see 'Edit on experimental page' button in 'Edit product' v1 page
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is disabled
WHEN I open a 'Edit Product' BO page
THEN I cannot see the 'Edit on experimental page' button
=> :heavy_check_mark: 

Scenario: Redirect user from v2 page to v1 page
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is disabled
WHEN I attempt to visit the BO page v2 using the URL /admin-dev/index.php/sell/catalog/products-v2/{PRODUCT-ID}/edit?_token={TOKEN}
THEN I am redirected to BO page v1
AND a flash message tells me why I was redirected
=> :heavy_check_mark: 

Scenario: Can see 'New product on experimental page' button in Product Catalog
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is enabled
THEN on the BO listing page 'Product Catalog' I can see the 'New product on experimental page' button
=> :heavy_check_mark: 

Scenario: Can see 'Edit on experimental page' button in Product Catalog
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is enabled
WHEN I open the Product Catalog BO page and I click on the dots menu on the right of a product
THEN I can see a link to Page v2 with name 'Edit on experimental page'
=> :heavy_check_mark: 

Scenario: Can see 'Edit on experimental page' button in 'Edit a product' v1 page
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is enabled
WHEN I open a 'Edit product' BO page
THEN I can see the 'Edit on experimental page' button
=> :no_entry_sign:  : OK When on Edit a product page V1, but when on Add a Product V1 : The message should be "Add on experimental page" : Since this is a trivial issue, it can be done after the merge on another regression PR in order not to block this PR and risk to miss the feature freeze.
Issue created : https://github.com/PrestaShop/PrestaShop/issues/24068

Scenario: Can see 'Back to standard page' button in Product Catalog
AS a merchant
WHEN I Install PrestaShop 1.7.8
PROVIDED the Product Page V2 feature flag is enabled
WHEN I open a Product Page V2 page
THEN I can see a link to Page v1 with name 'Back to standard page'
=> :no_entry_sign:  : The button in only displayed on Edit Product V2 page, not on Add Product V2 page : The behavior on Add Product V2 page was not specified and is not easy to implement (because we need to think what will happen if such a button is displayed : Should we save the data entered on V2 Page and then go to Edit Product V1 page ? Or should we lose the data and go to an "empty" Add Product V1 ?). Since this is more a non-specified behavior than a bug on this PR, I think it does not block the PR.
Issue created : https://github.com/PrestaShop/PrestaShop/issues/24071

**### Conclusion**

**Since the 2 "** :no_entry_sign: **" behavior are, in my opinion, not blocking, it's a QA** :heavy_check_mark: **!**

Thanks !

**Comment by Progi1984** [comment]
Thanks @matks & @Robin-Fischer-PS

**Pull request merged**
